### ISSUE TITLE
Presence docs to discourage building of presence sets

### ISSUE THREAD

**Issue by mattheworiordan** [ably/docs#2073]
A partner of ours has been building an app which has erratic presence behaviour. Upon inspecting the code, we see the same pattern of developers continuing to try and build their own presence sets based on `enter`, `leave`, `update` events, which is flawed and should never be done, they should instead rely on the presence `get` method to get the complete presence set. We suffered this same issue ourselves with our Spaces SDK development.  If our own internal staff are making these mistakes, and partners are making these mistakes, customers will certainly be making these mistakes.

Here is the sample code for ref:

![MO screenshot 2023-11-29 at 23 35 43](https://github.com/ably/docs/assets/43789/68087b1b-9747-460f-be0c-feeac1106bd6)

We really need to rethink our presence docs to encourage the right behaviours. When I read https://ably.com/docs/presence-occupancy/presence, I see no mention of why users should not handle enter, leave, update events and build their own presence set, and in fact the get method to get the presence set only appears a couple of (virtual) pages down the page.

┆Issue is synchronized with this [Jira Task](https://ably.atlassian.net/browse/EDU-1457) by [Unito](https://www.unito.io)

**Comment by mattheworiordan** [comment]
@paddybyers @SimonWoolf @owenpearson @ttypic I was just looking at [that presence code again](https://github.com/ably/demo/blob/8ed740179f9638b534903318694bf23c69c9fa6c/src/providers/members-provider.tsx#L30-L59), and it got me thinking about whether this is a documentation issue, or in fact an SDK / API issue. 

Right now, SDKs emit the raw protocol message presence events for `enter`, `leave`, `update` (and `sync`), which is where the problem lies. There is not really any good reason for us to emit these raw events from a consumer of the API's perspective. Our raw protocol messages are not all that useful to customers given they need to understand that messages can arrive out of order and they need to understand how to apply CRDT-like logic to get the intended result set. 

If we focus on what the customer would want instead of how this works internally, I would change things as following:
* We only ever emit enter, leave, and update events in a way that makes sense to the consuming user i.e. if they wrote code like above, it would work.
* We include an argument in the callback with a reference to the complete presence set anyway to encourage developers to use that when they need the entire presence set as opposed to knowing what has changed

WDYT?

**Comment by sacOO7** [comment]
This seems to be a react example. where they want to update the view when the dataset changes. 
More specifically, they want to get `updated` data set whenever any change to the dataset happens. 
I think `get` only returns a snapshot of members at a particular moment.
That said, `laravel-echo` devs exposed minimum right methods for presence. 

https://github.com/laravel/echo/blob/master/src/channel/presence-channel.ts#L10
 
In the Echo.join().here(members => {}) implementation, the callback is called with updated members a new member joins, updates, or leaves the channel.

**Comment by mattheworiordan** [comment]
> I think get only returns a snapshot of members at a particular moment.

Yup, why's that an issue?

> To expose this we had to update the code in ably laravel echo fork =>

Yup, and it just calls `presence.get` which IS the recommended behaviour because trying to build your own presence set from the events is painfully difficult and error prone.

**Comment by sacOO7** [comment]
Almost all of the mobile/web UI frameworks updates components based on state change, so we expose a method on `presence` channel like
```
channel.presence.onMembersUpdated( updatedMembers => {
})
```
instead of them doing 
```
// This can get complex based on each SDK api style
        this.channel.presence.subscribe(['enter', 'update', 'leave'], () =>
            this.channel.presence.get((err, members) =>
                callback(members.map(({data}) => data), err)
            )
        );
```

**Comment by sacOO7** [comment]
Yeah, introducing any breaking/non-breaking change that can make it easier for users to access changed members 👍

**Comment by SimonWoolf** [comment]
We do document this: https://ably.com/docs/best-practice-guide#use-existing-presence-set . I know because I wrote most of that entry. The problem is that it's in the best practice guide, which is hard to discover and no-one ever reads. The solution is to move it into the normal presence docs, which is a change that was agreed some time ago.

> We only ever emit enter, leave, and update events in a way that makes sense to the consuming user i.e. if they wrote code like above, it would work.

The main reason the above code is broken is that it assumes the presence set is uniquified by clientId, which it isn't, it's uniquified by clientId+connectionId. This is not a thing we can change just with in the client library api change, it would require a completely different presence model, client and serverside. I don't think this is a change it would be sensible to make at this point, to put it mildly.

> We include an argument in the callback with a reference to the complete presence set anyway to encourage developers to use that when they need the entire presence set as opposed to knowing what has changed

Sure, no objection to this. Trivial and backwards-compatible change for ably-js.

**Comment by mattheworiordan** [comment]
> The main reason the above code is broken is that it assumes the presence set is uniquified by clientId, which it isn't, it's uniquified by clientId+connectionId

That's the main reason, but not the only reason. The former is easy to explain to a developers and they can reason about (users can be present more than once with different connection IDs), but the latter is much harder (enter events don't necessarily mean the user has entered because a leave event can arrive after but with an earlier timestamp). I think we should fix that latter problem in our SDK API because emitting the raw protocol events is not the API a developer would expect to see. We should have an abstraction over our underlying protocol instead of exposing it. That's not a big change (and requires not realtime system changes), but would materially reduce the likelihood of developers mistakenly using this API. WDYT?

**Comment by mattheworiordan** [comment]
Apologies for the noise in this issue, I've corrected [my first follow on comment](https://github.com/ably/docs/issues/2073#issuecomment-1839135830)

Saying all that, I do think we should:
* Improve the docs, as mentioned by @SimonWoolf 
* We should include the presence set in the presence event callback, as suggested in https://github.com/ably/docs/issues/2073#issuecomment-1839251011

**Comment by sacOO7** [comment]
Link to internal discussion ->  https://ably-real-time.slack.com/archives/CURL4U2FP/p1701711106209279

**Comment by sacOO7** [comment]
They would need a callback to be executed everytime it's updated. Just exposing `members` as a property won't notify the change : (
